core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala

```diff
@@ -0,0 +1,293 @@
+  }
+
+  // Visible for testing.
+  private[dynalloc] def isExecutorIdle(id: String): Boolean = {
+    Option(executors.get(id)).map(_.isIdle).getOrElse(throw new NoSuchElementException(id))
+  }
+
+  // Visible for testing
+  private[dynalloc] def timedOutExecutors(when: Long): Seq[String] = {
+    executors.asScala.flatMap { case (id, tracker) =>
+      if (tracker.timeoutAt <= when) Some(id) else None
+    }.toSeq
+  }
+
+  // Visible for testing
+  def executorsPendingToRemove(): Set[String] = {
+    executors.asScala.filter { case (_, exec) => exec.pendingRemoval }.keys.toSet
+  }
+
+  /**
+   * This method should be used when updating executor state. It guards against a race condition in
+   * which the `SparkListenerTaskStart` event is posted before the `SparkListenerBlockManagerAdded`
+   * event, which is possible because these events are posted in different threads. (see SPARK-4951)
+   */
+  private def ensureExecutorIsTracked(id: String): Tracker = {
+    executors.computeIfAbsent(id, _ => new Tracker())
+  }
+
+  private def updateNextTimeout(newValue: Long): Unit = {
+    while (true) {
+      val current = nextTimeout.get()
+      if (newValue >= current || nextTimeout.compareAndSet(current, newValue)) {
+        return
+      }
+    }
+  }
+
+  private class Tracker {
+    @volatile var timeoutAt: Long = Long.MaxValue
+
+    // Tracks whether this executor is thought to be timed out. It's used to detect when the list
+    // of timed out executors needs to be updated due to the executor's state changing.
+    @volatile var timedOut: Boolean = false
+
+    var pendingRemoval: Boolean = false
+
+    private var idleStart: Long = -1
+    private var runningTasks: Int = 0
+
+    // Maps RDD IDs to the partition IDs stored in the executor.
+    // This should only be used in the event thread.
+    val cachedBlocks = new mutable.HashMap[Int, mutable.BitSet]()
+
+    // For testing.
+    def isIdle: Boolean = idleStart >= 0
+
+    def updateRunningTasks(delta: Int): Unit = {
+      runningTasks = math.max(0, runningTasks + delta)
+      idleStart = if (runningTasks == 0) clock.getTimeMillis() else -1L
+      updateTimeout()
+    }
+
+    def updateTimeout(): Unit = {
+      val oldDeadline = timeoutAt
+      val newDeadline = if (idleStart >= 0) {
+        idleStart + (if (cachedBlocks.nonEmpty) storageTimeoutMs else idleTimeoutMs)
+      } else {
+        Long.MaxValue
+      }
+
+      // Update the timeout before checking whether a recomputation is needed. This ensures
+      // that if the EAM sees the new timeout when double-checking whether an executor is really
+      // timed out, after setting the "timedOut" flag checked below.
+      timeoutAt = newDeadline
+
+      // If the executor was thought to be timed out, but the new deadline is later than the
+      // old one, ask the EAM thread to update the list of timed out executors.
+      if (timedOut && newDeadline > oldDeadline) {
+        nextTimeout.set(Long.MinValue)
```

Comment:
Why set current `nextTimeout` to `MinValue` ? Do you mean we need eagerly recompute timed out executors at next round to `update the list of timed out executors` ? But, as far as I know from the code, each time EAM thread invoke on `timedOutExecutors()`, it gets the list of timed out executors and asks cluster manager to kill them by `ExecutorAllocationClient` immediately. So, if an executor has already been considered timed out and killing after last round, will there be any difference for that executor if we eagerly `update the list of timed out executors` in next round ?

I guess, here, you mean to handle the case that an event comes making the executor to be non timed out again after the  executor just considered to be timeout nearly **after scanning**. But, as I mentioned above, I think it won't work.  Instead of **after scanning**, I think, maybe, we can do optimization for the case when it happens **during scanning**.  For example, wrapping a `while` loop on scanning and exit until all timed out executors' `timeOut` to be `true`:

```
// this check guard against event which updates deadline during scanning
while(!checkTimedOut(timedOutExecs)) {
 ...
 timedOutExecs = scanning {
 }
 ...
}
```

WDYT ?

Reply:
> it gets the list of timed out executors and asks cluster manager to kill them by ExecutorAllocationClient immediately. So, if an executor has already been considered timed out and killing after last round, will there be any difference for that executor if we eagerly update the list of timed out executors in next round ?

only the CoarseGrainedSchedulerBackend has a consistent view of which executors are actually running tasks, and so it has ultimate say in whether or not to kill an executor.  You might not have killed an executor earlier because it was actually busy.

So here is where you "catch up" with the view of the CGSB, so you know that executor shouldn't be considered timed out and requested to be killed next time around.

(at least I think so, marcelo can confirm ...)

Reply:
What Imran said, with a complication.

The scenario here is decidedly rare. The EAM thread sees a timed out executor at the same time a new task is starting on it. So the EAM tries to kill it, but the CGSB says no.

While that's happening the task start event arrives, and the monitor now knows the executor is not idle anymore. But the executor is still there in the list of timed out executors, so on every cycle, the EAM would still try to kill it and the CGSB will keep saying no. Until it becomes idle again (but not timed out!), and then it will be killed, but at the wrong time.

So it needs to be removed from the timed out list to avoid that problem.

There's still a case where the task start event isn't processed until after the task has actually ended, and the EAM thread runs again and actually manages to kill an executor when it should. But that should be even rarer, and I'm not sure that can be fixed, since it involves synchronization with threads that this code has no control over.

Reply:
Make sense.@squito @vanzin

> But the executor is still there in the list of timed out executors, so on every cycle, the EAM would still try to kill it and the CGSB will keep saying no. 

I just realized that we have a cached `timedOutExecs` list in monitor. But there's one last point we may overlook. Once an executor has considered to be timed out in a previous scanning round, its `pendingRemoval` would be marked as `true`. So, in next scanning round, we'll filter it out before checking its `deadline`. Then, it won't appear in `timedOutExecs` due to `pendingRemoval` rather than its new `deadline`. We should cover this.

But, why we need to cache `timedOutExecs` in monitor ? Could we always return a latest timed out execs list to EAM on every cycle(e.g. return empty when now < nextTimeout.get) ?

Reply:
> So, in next scanning round, we'll filter it out before checking its deadline. Then, it won't appear in timedOutExecs due to pendingRemoval rather than its new deadline. We should cover this.

I'm not sure I understand this part.  `pendingRemoval` is after the CGSB has acknowledged that it will remove the executor.  We shouldn't need to send any more messages about killing the executor after this.

> But, why we need to cache timedOutExecs in monitor ? Could we always return a latest timed out execs list to EAM on every cycle(e.g. return empty when now < nextTimeout.get) ?

I think you might be right about this, I would need to read it more carefully ... but in any case, I think that would make it significantly harder to follow.  Its a lot easier to reason about if `ExecutorMonitor.timedOutExecutors()` consistently returns the right list.

Reply:
> pendingRemoval is after the CGSB has acknowledged that it will remove the executor. We shouldn't need to send any more messages about killing the executor after this.

I revisit the code, you @squito are right. Thanks for pointing it out.

Reply:
> But, why we need to cache timedOutExecs in monitor ?

As explained the a comment where the `timedOutExecs` variable is declared, that's to avoid scanning the whole list of executors every 100ms.

Reply:
> that's to avoid scanning the whole list of executors every 100ms.

This doesn't make sense. Take for a not appropriate example, if we always got `now >= nextTimeout.get()`, then, we have to scanning every 100ms. I mean, that, we can't avoid `scanning the whole list of executors every 100ms` just by the cached `timedOutExecs`.  We can avoid scanning if and only if `now < nextTimeout.get()` and we can just return empty list to EAM.

Reply:
Hmmm, possibly. I think after some recent changes were made to the way I'm using `nextTimeout` it can replace the cached list. Let me take a look.

Reply:
Yeah, this doesn't work because of an edge case.

When calculating the timed out executors, the next timeout is set for when the next /non-timed/ out executor will time out. That is the important part.

Now imagine that you return 3 timed out executors and the EAM doesn't kill any of them (either because of limits, or because the CGSB doesn't do it). Now you will not get a call to `executorsKilled`, so the next timeout will not be reset.

Which means that next time you call `timedOutExecutors()` it will return an empty list, instead of the 3 timed out executors.

So you need that cache.

Reply:
BTW you could say "just call `executorsKilled` with an empty list" but then that means the optimization doesn't work, and you're basically calculating the timed out list on every EAM schedule call.

Reply:
Okay, I see the problem here. But, things here I care about is that, with a cached list, EAM may performs unnecessary executors removing for many many times before nextTimeout change. For example, we have 3 executors in monitor, and EAM gets those 3 timeout executors in a single round scanning. That's means, we can not update nextTimeout during this scanning and nextTimeout would just be `Long.MaxValue` before new events(which updates nextTimeout) comes. And, for EAM, before nextTimeout updates, it always gets those 3 cached timed out executors on every 100ms, and performs unnecessary executor removing again and again. This case may be rare, or have less impact on EAM,  but I do worry about this similar possible unnecessary behavior  with that cache list.

I think remove that cache list can be achieved with a little more change:

```
if (deadline > now) {
 newNextTimeout = math.min(newNextTimeout, deadline)
 exec.timedOut = false
 false
} else {
 exec.timedOut = true
 true
}
```

1. instead of only updating `newNextTimeout` for non timed out executors, we should update it for timed out executors, too

2. when we get a call to `executorsKilled(executorIds)`,  updates `nextTimeout` comparing to those killed executors' `timeoutAt`, rather than just set to Long.MinValue

Reply:
> with a cached list, EAM may performs unnecessary executors removing for many many times before nextTimeout change

That's actually the point. If the EAM did not kill the executors, it most probably means that there's an event that will be delivered to the listener and cause the next timeout to be updated (e.g. a task start).

If such event doesn't arrive, then the executors are still timed out, and will be always returned to the EAM. Then it may just decide not to kill them again (e.g. because the lower limit of executors has been reached).

All that is fine, and cheap computationally.

> instead of only updating newNextTimeout for non timed out executors, we should update it for timed out executors, too

That means that if no executors are killed, you'll be re-scanning the whole executor list on every EAM interval until something changes, which is the exact thing the cache is trying to avoid.

Reply:
Make sense.